### ISSUE TITLE
Confusing documentation of mouse left/right click

### ISSUE THREAD

**Issue by dngulin** [minetest/minetest#2838]
Game description about it:

```
- Mouse left: dig/punch
- Mouse right: place/use
```

But item usage triggered by `left mouse button`, this is very strange behavior. For example, when i first try to gather apples, i get a problem: instead of gather apples my character eat gathered apple and then gathering a next apple. One apple at beginning and one apple at end of gathering :)

Way to solve this trouble: change item usage button to `right mouse button`, as written in controls description. (Or change current item at gathering. But i can't know what will be triggered by left mouse button for random item: usage or punch. This is bad behavior.)

Another trouble is a usage conflict. Only one button for two different usages: `selected item usage` (eat apple, write book, etc) and `node interaction` (open/close door, open chest, change a sign text, etc).

Minecraft-like way to solve this trouble: add `Shift` key modifier. In conflict state: `right mouse button` = node interaction, `Shift + right mouse button` = self item usage or place node. For minetest this is not best behavior, beacuse `Shift` also used for down the ladder (modifier usage conflict).

I think, more true way to solve usage conflict: separate different action to different control keys:
- `Left mouse button`: dig/punch
- `Right mouse button`: current item use/place node
- `E`: node/entity interaction (usage button in settings)

May be, this also need to change lua_api: rename `on_rightclick` to `on_interaction`.

PS. Sorry for my bad English.

**Comment by C1ffisme** [comment]
Yeah, we need to make eating use the right mouse button. And we need an eating animation.

It's pretty bad game design when you eat an apple as you are collecting them. As a temporary workaround though, put apples in the back of your inventory and then mine them.

**Comment by est31** [comment]
Just select another item while collecting the apples. You can't collect stone either, because you'll have to select a pick for it. For me, this is a gameplay element.

**Comment by figec** [comment]
> Just select another item while collecting the apples. You can't collect stone either, because you'll have to select a pick for it. For me, this is a gameplay element.

In this case need to change the description of controls to:
```
- Mouse left: dig/punch/use item
- Mouse right: place/use
```

**Comment by figec** [comment]
Also, may be need to add indicator for items that can be used.

**Comment by wischi-chr** [comment]
+1 for seperate use-button (as suggested: E) most games handle use/action that way because it's way more intuitive.

**Comment by paramat** [comment]
:-1:

**Comment by 0-afflatus** [comment]
It's not really up to the engine to determine how the buttons are used is it?
There is also another PR about making other keys assignable (sorry, can't be bothered to find the reference right now), which would seem to obviate the need for this.

**Comment by paramat** [comment]
More a subgame issue that can be solved in apple code or wherever.

**Comment by Ferk** [comment]
There's still a sort of inconsistent naming problem.

The engine defines "on_use" as the callback when "left clicking", yet the same engine indeed defines left click for "dig/punch", while right click for "place/use":

https://github.com/minetest/minetest/blob/bd6b90359c654f4c75964755e476a8bfd90114ba/src/game.cpp#L4649-L4650

It's very confusing how the "on_use" callback doesn't correspond with what the user is told "use" means.

I propose establishing a fixed terminology and stick with it to avoid confusion. Something like:

    - Mouse left:  use
    - Mouse right: interact

 Then make clear that the default use of an item/node is punching with it, and that the default interaction is placing a node. Then mods and games can change what an item does when punching or when interacting with it.

An important difference that the engine makes between these 2 actions is that the "on_use" (left click) is **always** executed, whenever the button is pressed, whereas the "on_righclick" ("interact") is executed **only when a node in the environment is pointed and on reach**.

This means it is bad design for a game or mod to use "on_rightclick" for any action that is not meant to be an interaction with a node in the environment. If you use "on_rightclick" to feed on an apple, the user will be confused when he is trying to eat the apple while pointing to the sky or the horizon and he is not able to do it.

**Comment by paramat** [comment]
Ok will reopen as a documentation issue.

**Comment by Fixer-007** [comment]
>But item usage triggered by left mouse button, this is very strange behavior. For example, when i first try to gather apples, i get a problem: instead of gather apples my character eat gathered apple and then gathering a next apple. One apple at beginning and one apple at end of gathering :)

This bothered me since ever. Do not close this issue, maybe it needs deeper thinking.

**Comment by Ferk** [comment]
@Fixer-007 yes, the problem is that everyone is used to Minecraft-like controls, where you can punch and destroy nodes with anything, even with a string or an apple. It doesn't help that having an "on_use" is not very common in Minetest, unless it's some sort of special tool, or food.

At the beginning I also didn't like it but I can see the rationale of not using rightclick, because it takes away the possibility of placing them as a node, like apples. Though maybe a combination of sneak+rightclick could have been used to force the placing of the apple (which is also used to place items even when pointing to nodes like the furnace or chest).

It might be too late to change this but I guess the initial idea might have been that players should only use the right tools for the right things, and most items should not allow punching at all, only collect apples with your hands or with some tool that can collect them, mine with pickaxes, chop trees with axes, etc. (some stuff could have been added to make rudimentary tools out what you can gather with your hands). It wouldn't be so strange that clicking with an apple doesn't make you punch if people learn that only some items allow you to punch with them.

Another thing that doesn't help is that the punch animation is hardcoded by the engine, the same punch animation is done whether you press left or right click and there's no way to press them without the animation happening. An attempt was done in #4297 to add custom animation to the items when punching, so it would be possible to make a movement with an apple to make it look like you are biting it, to a book when you are reading it, etc.

**Comment by ixPost** [comment]
> `E`: node/entity interaction (usage button in settings)

It won't work because on Liberty Land & Your Land when you press J and then walk you will walk at normal speed but if you hold down E you'll run. So if we want to implement

> `E`: node/entity interaction (usage button in settings)

then to run players will have to press J every time they want to run and press it again when they want to walk, effectively a similar thing will happen as when you leave the Caps Lock on by accident.

**Comment by Wuzzy2** [comment]
Part of this is a MTG issue because eating apples is mapped to LMB. In theory, you could map it to RMB tho. MCL2 actually does this. You eat with RMB in MCL2.
If you think eating apples should not be mapped to LMB, post an issue in the MTG repo. I wonder if this could be classified as a bug. Anyway …

Y'all thinkin with a MTG brain again, turn it off. :D We need to look at this more generally.

Let's analyze this bit for bit:

> - Mouse left: dig/punch

This is definitely true. Leftclick is digging and punching by default. This is true for pretty much all games that allow digging and punching. I don't know any game that awkwardly maps digging and punching to another key.

The problem here is that the word "use" is missing although it has an obvious "use" action in MTG (i.e. eating).

> - Mouse right: place/use

The "place" part is also correct.

The "use" part is the one that causes problems.
First, the "use" is not entirely incorrect. "Use" not only refers to your wielded item, but to pointed nodes. You open the node's formspec with this key. You trigger `on_rightclick` action of a node or entity. This is what the word "use" is referring to.

It is also possible that items have an `on_rightclick` action, just not in MTG. MCL2 does this extensively, so MCL2 ironically is truer to Minetest than MTG. :D

The funny thing is, the Esc help is wrong in MTG but correct in MCL2. Therefore, I don't think there is a solution just by simply rewriting the Esc text, it will always be wrong in some game. Changing MTG alone is also no real solution as many MTG forks still exist and you can't automatically change them as well.

We also have to distinguish between "item use" and "pointed thing use". This is not the same. The help should probably reflect that.

## The suggestion

Allow games to rewrite the Esc control help by assigning a different description for each key, or by hiding a key entirely.
But allow them to only rewrite the action like "dig/punch", not the key name itself (that one should still be provided by the engine).
I am not proposing a completely freeform text approach.

Another benefit is that games can document special uses of the Aux1 key.

If you think about it, this list has a LOT of implications about games that might not be true for many games.
Not all games allow digging or placing.
Not all games allow dropping items.
Not all games allow jumping or walking.
And so on ...
Especially with all the Game Jam games this should be more than obvious now.
I think the problem here is a bit broader than just "LMB = eating".


On the other hand, I still think the default Esc text should still be reasonable. I suggest the broadest possible interpretation:

> - Mouse left: dig/punch/use
> - Mouse right: place/use

The change here is by adding "use" to Mouse left.
The "use" in both cases could refer to either using the pointed thing or the wielded item. While this does not apply to all games, the point here is that it COULD be true in the game. Also, the meaning of "use" is deliberately kept vague here. I think this generic approach might be the best short-term solution. That would fix the problem with all those "LMB = eat" games.

And if a game wants to be more specific, they can use the overriding.

Summary:

1. Allow games to overwrite action descriptions on a case-by-case basis
2. Add the word "use" to "Mouse left" in the default controls help

**Comment by Ferk** [comment]
> Mouse right: place/use

To me, the problem is in the wording. 

Calling an action "use" is confusing because "use" can mean many things when you don't know what it refers to (the held item? the pointed thing?).

This is why long ago I advocated to calling it "interact" when the thing "used" was the pointed thing. But it never caught on... and nowadays there's `on_secondary_use`, making it more confusing.

In the past the right click was meant to be used as an "interact with the pointed thing while carrying the item" button, it was not really reliable to use right click as a "use the item" button because the callback function only got called when there was something being pointed at (looking at the sky and pressing right click with an item did not trigger the "secondary use"). This used to be intentional, the more Minetest-way (vs Minecraft) to differentiate the use of left click and right click controls, and it's the reason things like apples used left click.

Left click was for actions that didn't need to point to something to apply them, so it was more of a "use the item by itself", while right click used to be meant for actions meant for acting upon the thing being pointed in some way (even if it involved the item being held), so a sort of "interact with enviroment while holding the item". The interaction might depend on what item was held and what thing was pointed at. Sometimes the interaction meant placing the item as a node.

But nowadays the functionality was changed so it no longer works that way. It really is just a "secondary use" that instead of defaulting to punching, happens to sometimes trigger methods in the nodes pointed at.
